### ISSUE TITLE
merging structs with opiton "WithOverride" not consistent

### ISSUE THREAD

**Issue by ghouscht** [imdario/mergo#54]
Hi,
In my opinion merging nested structs isn't really consistent when using the option "WithOverride". If you nest a struct "by value" or "by pointer" the behaviour of merge with override is different. (Possibly a duplicate of #14)
Example Code:
```golang
package main
import (
        "fmt"
        "log"

        "github.com/imdario/mergo"
)

type PtrTest struct {
        Str    string
        Number int
        Nested *Nested
}

type Test struct {
        Str    string
        Number int
        Nested Nested
}

type Nested struct {
        NestedStr    string
        NestedNumber int
}

func main() {
        dstPtr := &PtrTest{
                Str: "This is a non empty test string",
                Nested: &Nested{
                        NestedStr:    "This is a non empty string in a nested struct",
                        NestedNumber: 42,
                },
        }

        srcPtr := &PtrTest{
                Str:    "This is also a string",
                Number: 4,
                Nested: &Nested{},
        }

        if err := mergo.Merge(dstPtr, srcPtr, mergo.WithOverride); err != nil {
                log.Fatal(err)
        }
        fmt.Printf("Result with 'Nested' as pointer: %v\n", dstPtr.Nested)

        dst := &Test{
                Str: "This is a non empty test string",
                Nested: Nested{
                        NestedStr:    "This is a non empty string in a nested struct",
                        NestedNumber: 42,
                },
        }

        src := &Test{
                Str:    "This is also a string",
                Number: 4,
                Nested: Nested{},
        }

        if err := mergo.Merge(dst, src, mergo.WithOverride); err != nil {
                log.Fatal(err)
        }
        fmt.Printf("Result with 'Nested' as value: %v\n", dst.Nested)
}
```
This code prints the following two lines:
```
Result with 'Nested' as pointer: &{ 0}
Result with 'Nested' as value: {This is a non empty string in a nested struct 42}
```

Quote from the documentation: _WithOverride will make merge override non-empty dst attributes with non-empty src attributes values._

I expected that behaviour also from nested structs (by pointer). Maybe this is just a misunderstanding of mine, because in my example the pointer to the nested struct isn't empty, but the element to which the pointer points is. 
Actually if you remove the pointer to the empty element the example works like I'd expected it. However I got a use case which this solution isn't really applicable :smiley:.

Without looking at your code I think it should be possible to make a consistent use of "WithOverride" possible (also with my example). However I think it won't be very easy and maybe adds a bit of complexity. 
If there is interest I can take a look at it and try to implement it. If not I think it should at least be documented for other people.

Please let me know your opinion about that.
Thanks for your work and regards, 
Thomas

**Comment by ghouscht** [comment]
Wrote a few lines of code to solve my problem (see referenced commit). Still some work needed as stated by the TODOs in the code. If there is interest in this, I can continue with solving the TODOs and writing some more tests.

Currently there is one failing test with my additions (TestIssue23MergeWithOverwrite) which I also need to further investigate.

Please let me know wheter I should continue or not.
- Thomas

**Comment by imdario** [comment]
I'll check this later. Thanks!

**Comment by ghouscht** [comment]
The changes and tests from the referenced commit were actually wrong (had some troubles with the terminology of "WithOverride").

Fixed the issues a few minutes ago and https://github.com/ghouscht/mergo  branch master should be ok now. I also found out why "TestIssue23MergeWithOverwrite" was failing with my changes, see commit https://github.com/ghouscht/mergo/commit/f2027b273c442aaa4b35f14e022a561d0ea69b60 for more details.

My code should be ok now but needs very careful review to make sure we do not brake things out there...

**Comment by ghouscht** [comment]
any news on this?

**Comment by imdario** [comment]
Please, make a PR. It will be easier for me to check your changes.

**Comment by ghouscht** [comment]
Opened PR #62 for you, thanks for checking.

**Comment by imdario** [comment]
It took some time but I'm checking it now. I'm having problems to understand the full reach of your modification because the PR has smaller edits to unrelated things. Anyway, I'll do my best and merge as soon as possible.

**Comment by imdario** [comment]
Ok, I think now I get it. Sorry, I can't accept it as is. It doesn't feel safe to change behavior in such way.

Also, I ran your example code and I must say that it works as expected. Please follow my explanation in #52:

[...] As per MergeWithOverwrite's documentation:

> MergeWithOverwrite will do the same as Merge except that **non-empty dst attributes will be overriden by non-empty src attribute values**.

[...] time.Time can be semantically empty (as in IsZero() returns true) but it is a struct. Structs in Go don't have a zero value by themselves, as described in the [official documentation about zero value](https://golang.org/ref/spec#The_zero_value):

> When storage is allocated for a variable, either through a declaration or a call of new, or when a new value is created, either through a composite literal or a call of make, and no explicit initialization is provided, the variable or value is given a default value. Each element of such a variable or value is set to the zero value for its type: false for booleans, 0 for integers, 0.0 for floats, "" for strings, and nil for pointers, functions, interfaces, slices, channels, and maps. This initialization is done recursively, so for instance each element of an array of structs will have its fields zeroed if no value is specified.

In other words, an empty struct isn't an empty value per se.

I'll accept your PR but I will merge it removing your edits in merge.go.
